File: tl/src/main/java/com/github/badoualy/telegram/tl/api/TLMessageActionPaymentSent.java
package com.github.badoualy.telegram.tl.api;

import com.github.badoualy.telegram.tl.TLContext;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

import static com.github.badoualy.telegram.tl.StreamUtils.readLong;
import static com.github.badoualy.telegram.tl.StreamUtils.readTLString;
import static com.github.badoualy.telegram.tl.StreamUtils.writeLong;
import static com.github.badoualy.telegram.tl.StreamUtils.writeString;
import static com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_CONSTRUCTOR_ID;
import static com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_INT64;
import static com.github.badoualy.telegram.tl.TLObjectUtils.computeTLStringSerializedSize;

/**
 * @author Yannick Badoual yann.badoual@gmail.com
 * @see <a href="http://github.com/badoualy/kotlogram">http://github.com/badoualy/kotlogram</a>
 */
public class TLMessageActionPaymentSent extends TLAbsMessageAction {

    public static final int CONSTRUCTOR_ID = 0x40699cd0;

    protected String currency;

    protected long totalAmount;

    private final String _constructor = "messageActionPaymentSent#40699cd0";

    public TLMessageActionPaymentSent() {
    }

    public TLMessageActionPaymentSent(String currency, long totalAmount) {
        this.currency = currency;
        this.totalAmount = totalAmount;
    }

    @Override
    public void serializeBody(OutputStream stream) throws IOException {
        writeString(currency, stream);
        writeLong(totalAmount, stream);
    }

    @Override
    @SuppressWarnings({"unchecked", "SimplifiableConditionalExpression"})
    public void deserializeBody(InputStream stream, TLContext context) throws IOException {
        currency = readTLString(stream);
        totalAmount = readLong(stream);
    }

    @Override
    public int computeSerializedSize() {
        int size = SIZE_CONSTRUCTOR_ID;
        size += computeTLStringSerializedSize(currency);
        size += SIZE_INT64;
        return size;
    }

    @Override
    public String toString() {
        return _constructor;
    }

    @Override
    public int getConstructorId() {
        return CONSTRUCTOR_ID;
    }

    public String getCurrency() {
        return currency;
    }

    public void setCurrency(String currency) {
        this.currency = currency;
    }

    public long getTotalAmount() {
        return totalAmount;
    }

    public void setTotalAmount(long totalAmount) {
        this.totalAmount = totalAmount;
    }
}

File: tl/src/main/kotlin/com/github/badoualy/telegram/tl/api/TLMessageActionPaymentSent.kt
package com.github.badoualy.telegram.tl.api

import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_CONSTRUCTOR_ID
import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_INT64
import com.github.badoualy.telegram.tl.TLObjectUtils.computeTLStringSerializedSize
import com.github.badoualy.telegram.tl.serialization.TLDeserializer
import com.github.badoualy.telegram.tl.serialization.TLSerializer
import java.io.IOException

/**
 * messageActionPaymentSent#40699cd0
 *
 * @author Yannick Badoual yann.badoual@gmail.com
 * @see <a href="http://github.com/badoualy/kotlogram">http://github.com/badoualy/kotlogram</a>
 */
class TLMessageActionPaymentSent() : TLAbsMessageAction() {
    var currency: String = ""

    var totalAmount: Long = 0L

    private val _constructor: String = "messageActionPaymentSent#40699cd0"

    override val constructorId: Int = CONSTRUCTOR_ID

    constructor(currency: String, totalAmount: Long) : this() {
        this.currency = currency
        this.totalAmount = totalAmount
    }

    @Throws(IOException::class)
    override fun serializeBody(tlSerializer: TLSerializer) = with (tlSerializer)  {
        writeString(currency)
        writeLong(totalAmount)
    }

    @Throws(IOException::class)
    override fun deserializeBody(tlDeserializer: TLDeserializer) = with (tlDeserializer)  {
        currency = readString()
        totalAmount = readLong()
    }

    override fun computeSerializedSize(): Int {
        var size = SIZE_CONSTRUCTOR_ID
        size += computeTLStringSerializedSize(currency)
        size += SIZE_INT64
        return size
    }

    override fun toString() = _constructor

    override fun equals(other: Any?): Boolean {
        if (other !is TLMessageActionPaymentSent) return false
        if (other === this) return true

        return currency == other.currency
                && totalAmount == other.totalAmount
    }
    companion object  {
        const val CONSTRUCTOR_ID: Int = 0x40699cd0.toInt()
    }
}
